src/tool/subcommands/state_migration_cmd.rs

```diff
@@ -1,140 +1,104 @@
+                    .get(url.clone())
+                    .send()
+                    .await?
+                    .error_for_status()?;
+                let bytes = response.bytes().await?;
+                let car = CarStream::new(Cursor::new(bytes)).await?;
+                ensure!(car.header.version == 1);
+                ensure!(car.header.roots.len() == 1);
+                ensure!(&car.header.roots[0] == root);
+                anyhow::Ok((root, car.try_collect::<Vec<_>>().await?))
+            },
         )
-        .await?;
-
-    Ok(())
-}
-
-async fn download_bundle_if_needed(root: &Cid, url: &Url) -> anyhow::Result<PathBuf> {
-    if !ACTOR_BUNDLE_CACHE_DIR.is_dir() {
-        fs::create_dir_all(ACTOR_BUNDLE_CACHE_DIR.as_path())?;
-    }
-    let cached_car_path = ACTOR_BUNDLE_CACHE_DIR.join(format!("{root}.car"));
-    if cached_car_path.is_file() {
-        if let Ok(file) = async_fs::File::open(&cached_car_path).await {
-            if let Ok(true) = is_bundle_valid(root, BufReader::new(file)).await {
-                return Ok(cached_car_path);
-            }
-        }
+        .try_collect::<Vec<_>>()
+        .await?
+        .into_iter()
+        .unzip::<_, _, Vec<&Cid>, Vec<_>>();
+
+    let roots = roots.into_iter().cloned().collect::<CidHashSet>();
+    let blocks = blocks.into_iter().flatten().try_fold(
+        CidHashMap::new(),
+        |mut acc, Block { cid, data }| {
+            // TODO(aatifsyed): check cid matches data?
+            match acc.entry(cid) {
+                CidHashMapEntry::Vacant(v) => {
+                    v.insert(data);
+                }
+                CidHashMapEntry::Occupied(o) if o.get() == &data => {}
+                CidHashMapEntry::Occupied(_) => {
+                    bail!("clobbered data for block with cid {cid}")
+                }
+            };
+            anyhow::Ok(acc)
+        },
+    )?;
+
+    let mut car = vec![];
+
+    CarHeader {
+        roots: roots.into_iter().collect(),
+        version: 1,
     }
+    .write_stream_async(&mut car, &mut stream::iter(blocks))
+    .await?;
 
-    let tmp = tempfile::NamedTempFile::new_in(ACTOR_BUNDLE_CACHE_DIR.as_path())?.into_temp_path();
-    {
-        let response = global_http_client().get(url.clone()).send().await?;
-        let mut writer = BufWriter::new(async_fs::File::create(&tmp).await?);
-        futures::io::copy(
-            response
-                .bytes_stream()
-                .map_err(|reqwest_error| io::Error::new(io::ErrorKind::Other, reqwest_error))
-                .into_async_read(),
-            &mut writer,
-        )
-        .await?;
-        writer.flush().await?;
-        writer.close().await?;
-    }
-    if is_bundle_valid(root, BufReader::new(async_fs::File::open(&tmp).await?)).await? {
-        tmp.persist(&cached_car_path)?;
-        Ok(cached_car_path)
-    } else {
-        anyhow::bail!("Invalid bundle: {url}");
-    }
-}
+    let car_zst = zstd::encode_all(car.as_slice(), 3)?;
```

Comment:
This should reduce the file size by ~90%.

```suggestion
    let car_zst = zstd::encode_all(car.as_slice(), 17)?;
```